features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+    count: 16
+```
+
+# SDK API
+
+The following functions are all in the namespace `skygear.auth.mfa`.
+
+```typescript
+// Register TOTP
+
+interface CreateNewTOTPResult {
+  authenticatorID: string;
+  authenticatorType: "totp";
+  secret: string;
+}
+
+interface ActivateTOTPResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewTOTP(displayName?: string): Promise<CreateNewTOTPResult>;
+async function activateTOTP(authenticatorID: string, otp: string): Promise<ActivateTOTPResult>;
+
+// Register OOB
+
+interface CreateNewOOBOptions {
+  channel?: "sms";
+  phone: string;
+}
+
+interface CreateNewOOBResult {
+  authenticatorID: string;
+  authenticatorType: "oob";
+  channel: "sms";
+  phone: string;
+}
+
+interface ActivateOOBResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewOOB(options: CreateNewOOBOptions): Promise<CreateNewOOBResult>;
+async function triggerOOB(authenticatorID?: string): Promise<void>;
+async function activateOOB(authenticatorID: string, code: string): Promise<ActivateOOBResult>;
+
+// Error inspection
+
+function isMFARequiredError(err: unknown): boolean;
+
+// Authenticate
+
+interface AuthenticateExtraOptions {
+  requestBearerToken?: boolean;
+}
+
+interface AuthenticateResult {
+  bearerToken?: string;
+}
+
+interface TOTPAuthenticateOptions {
+  authenticatorID?: string;
+  otp: string;
+}
+
+async function authenticateWithTOTP(options: TOTPAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+interface OOBAuthenticateOptions {
+  authenticatorID?: string;
+  code: string;
+}
+
+async function authenticateWithOOB(options: OOBAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithRecoveryCode(code: string, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithBearerToken(token: string): Promise<void>;
+
+// Authenticator management
+
+type Authenticator = TOTPAuthenticator | OOBAuthenticator | RecoveryCodeAuthenticator | BearerTokenAuthenticator;
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  phone: string;
+}
+
+interface RecoveryCodeAuthenticator {
+  id: string;
+  type: "recovery_code";
+  createdAt: Date;
+}
+
+interface BearerTokenAuthenticator {
+  id: string;
+  type: "bearer_token";
+  createdAt: Date;
+  expireAt: Date;
+}
+
+interface RegenerateRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+async function listAuthenticators(): Promise<Authenticator[]>;
+async function deleteAuthenticator(authenticatorID: string): Promise<void>;
+async function regenerateRecoveryCode(): Promise<RegenerateRecoveryCodeResult>;
+```
+
+# Use Cases
+
+## Registration
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
```

Comment:
You mean `signupWithEmail`?

Reply:
I really meant `login`. The user should be able to sign normally. The access token they get will have `"amr": ["password"]`.

Reply:
I see, so the section title means MFA registration, not user registration.